### ISSUE TITLE
object_name_linter() misses S3 generics declared with `=` assignment

### ISSUE THREAD

**Issue by klmr** [r-lib/lintr#2507]
Consider the following code:

```r
f = function(x) {
  UseMethod("f")
}

g <- function(x) {
  UseMethod("f")
}

f.default <- function(x) {}

g.default <- function(x) {}
```

running `lintr::lint()` on this file with the default linters configuration gives the following output:

```
/path/test.r:1:3: style: [assignment_linter] Use <-, not =, for assignment.
f = function(x) {
  ^
/path/test.r:9:1: style: [object_name_linter] Variable and function name style should match snake_case or symbols.
f.default <- function(x) {}
^~~~~~~~~
```

The `object_name_linter` warning is unexpected: `f` is declared as a generic. In fact, directly calling `declared_s3_generics()` on the XML parse tree of the above code returns `"g"` where it should return `c("f", "g")`.

**Comment by MichaelChirico** [comment]
Nice find! This should be relatively easy to fix, are you interested in trying a PR? Should be a matter of including EQ_ASSIGN where only LEFT_ASSIGN is assumed now. There's also a relevant note in the wiki

**Comment by klmr** [comment]
Ah, I guess this note in the Wiki might address this:

> The outermost `<expr>` may be `<equal_assign>` or `<expr_or_assign_or_help>` instead.

**Comment by MichaelChirico** [comment]
> Ah, I guess this note in the Wiki might address this:
> 
> > The outermost `<expr>` may be `<equal_assign>` or `<expr_or_assign_or_help>` instead.

yep! it's possible we missed this in this case and only looked for <expr>
